core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/DefaultAwsS3V4HttpSigner.java

```diff
@@ -15,27 +15,202 @@
 
 package software.amazon.awssdk.http.auth.aws.internal;
 
-import software.amazon.awssdk.annotations.SdkInternalApi;
+import static software.amazon.awssdk.http.Header.CONTENT_LENGTH;
+import static software.amazon.awssdk.http.auth.internal.util.SignerConstant.STREAMING_UNSIGNED_PAYLOAD_TRAILER;
+import static software.amazon.awssdk.http.auth.internal.util.SignerConstant.UNSIGNED_PAYLOAD;
+import static software.amazon.awssdk.http.auth.internal.util.SignerConstant.X_AMZ_CONTENT_SHA256;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.ContentStreamProvider;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.auth.AwsV4HttpSigner;
+import software.amazon.awssdk.http.auth.SigV4RequestContext;
 import software.amazon.awssdk.http.auth.aws.AwsS3V4HttpSigner;
+import software.amazon.awssdk.http.auth.aws.internal.chunkedencoding.AwsS3V4ChunkSigner;
+import software.amazon.awssdk.http.auth.aws.internal.chunkedencoding.AwsSignedChunkedEncodingInputStream;
+import software.amazon.awssdk.http.auth.aws.internal.io.AwsChunkedEncodingConfig;
+import software.amazon.awssdk.http.auth.internal.AwsV4HttpProperties;
+import software.amazon.awssdk.http.auth.internal.checksums.ChecksumAlgorithm;
+import software.amazon.awssdk.http.auth.internal.checksums.ContentChecksum;
+import software.amazon.awssdk.http.auth.internal.checksums.SdkChecksum;
+import software.amazon.awssdk.http.auth.internal.util.CanonicalRequestV2;
 import software.amazon.awssdk.http.auth.spi.AsyncSignRequest;
 import software.amazon.awssdk.http.auth.spi.AsyncSignedRequest;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
 import software.amazon.awssdk.http.auth.spi.SyncSignRequest;
-import software.amazon.awssdk.http.auth.spi.SyncSignedRequest;
 import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.utils.StringUtils;
 
 /**
- * A default implementation of {@link AwsS3V4HttpSigner}.
+ * An implementation of a {@link AwsS3V4HttpSigner} for S3 use-cases, which includes chunked-encoded request payloads.
  */
-@SdkInternalApi
-public class DefaultAwsS3V4HttpSigner implements AwsS3V4HttpSigner {
+@SdkProtectedApi
+public final class DefaultAwsS3V4HttpSigner implements AwsV4HttpSigner<AwsS3V4HttpProperties> {
```

Comment:
This is a case where I did not separate out granular behavior into their own separate `AwsV4HttpSigner` implementations, like is done with query/presign/unsignedpayload.

Do we want separate impls here or is the collapsing fine, thoughts?

Reply:
Keeping them separate is fine.